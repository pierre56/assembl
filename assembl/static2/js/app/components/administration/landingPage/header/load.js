// @flow
import type { ApolloClient } from 'react-apollo';
import type { DatePickerValue } from './types.flow';
import DiscussionQuery from '../../../../graphql/DiscussionQuery.graphql';
import { convertEntriesToI18nValue, convertEntriesToI18nRichText, convertISO8601StringToDateTime } from '../../../form/utils';

export const load = async (client: ApolloClient, lang: string, fetchPolicy: FetchPolicy = 'cache-first') => {
  const { data } = await client.query({
    query: DiscussionQuery,
    fetchPolicy: fetchPolicy,
    variables: {
      lang: lang
    }
  });
  return data;
};

type Data = DiscussionQuery;

export function postLoadFormat(data: Data): DatePickerValue {
  return {
    headerTitle: convertEntriesToI18nValue(data.discussion.titleEntries),
    headerSubtitle: convertEntriesToI18nRichText(data.discussion.subtitleEntries),
    headerButtonLabel: convertEntriesToI18nValue(data.discussion.buttonLabelEntries),
    headerImage: data.discussion.headerImage,
    headerLogoImage: data.discussion.logoImage,
    headerStartDate: convertISO8601StringToDateTime(data.discussion.startDate),
    headerEndDate: convertISO8601StringToDateTime(data.discussion.endDate)
  };
}