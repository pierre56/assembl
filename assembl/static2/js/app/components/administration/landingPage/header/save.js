// @flow

import type { ApolloClient } from 'react-apollo';
import {
  createSave,
  convertToEntries,
  convertRichTextToVariables,
  getFileVariable,
  convertDateTimeToISO8601String
} from '../../../form/utils';
import type { DatePickerValue } from './types.flow';
import updateDiscussion from '../../../../graphql/mutations/updateDiscussion.graphql';

type UpdateDiscussion = updateDiscussion;

export const save = createSave('administration.landingPage.successSave');

const createVariablesFromValues = async (
  values: DatePickerValue,
  initialValues: DatePickerValue,
  client: ApolloClient
): Promise<UpdateDiscussion> => {
  const staticValues = {
    titleEntries: values.headerTitle ? convertToEntries(values.headerTitle) : null,
    buttonLabelEntries: values.headerButtonLabel ? convertToEntries(values.headerButtonLabel) : null,
    subtitleEntries: null,
    headerImage: getFileVariable(values.headerImage, initialValues.headerImage),
    logoImage: getFileVariable(values.headerLogoImage, initialValues.headerImage),
    startDate: convertDateTimeToISO8601String(values.headerStartDate),
    endDate: convertDateTimeToISO8601String(values.headerEndDate)
  };

  if (!values.headerSubtitle) return staticValues;

  const val = await convertRichTextToVariables(values.headerSubtitle, client);
  return {
    subtitleEntries: val || null,
    ...staticValues
  };
};

export const createMutationsPromises = (client: ApolloClient) => (
  values: UpdateDiscussion,
  initalValues: DatePickerValue
): Array<() => Promise<UpdateDiscussion>> => [
    () =>
      createVariablesFromValues(values, initalValues, client).then(vars =>
        client.mutate({
          mutation: updateDiscussion,
          variables: vars
        })
      )
  ];