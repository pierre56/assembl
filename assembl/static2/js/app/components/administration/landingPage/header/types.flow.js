// @flow

import type { FileValue, I18nValue, I18nRichTextValue, DatePickerInput } from '../../../form/types.flow';

export type DatePickerValue = {
  headerTitle: I18nValue,
  headerSubtitle: I18nRichTextValue,
  headerButtonLabel: I18nValue,
  headerImage: FileValue,
  headerLogoImage: FileValue,
  headerStartDate: DatePickerInput,
  headerEndDate: DatePickerInput
};

export type DatePickerValidation = {
  headerTitle: void | string,
  headerSubtitle: void | string,
  headerButtonLabel: void | string,
  headerImage: void | string,
  headerLogoImage: void | string,
  headerStartDate: void | string,
  headerEndDate: void | string
};