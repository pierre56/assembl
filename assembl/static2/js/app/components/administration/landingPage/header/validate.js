// @flow

import { I18n } from 'react-redux-i18n';
import { type DateTime } from '../../../form/types.flow';
import type { DatePickerValue, DatePickerValidation } from './types.flow';

export const validStartDate = (startDate: DateTime, endDate: DateTime): boolean => startDate <= endDate;

export const validEndDate = (startDate: DateTime, endDate: DateTime): boolean => endDate >= startDate;

export const validateDatePicker = (values: DatePickerValue): DatePickerValidation => ({
  headerTitle: undefined,
  headerSubtitle: undefined,
  headerButtonLabel: undefined,
  headerImage: undefined,
  headerLogoImage: undefined,
  headerStartDate: validStartDate(values.headerStartDate.time, values.headerEndDate.time)
    ? undefined
    : I18n.t('administration.landingPage.header.startDateError'),
  headerEndDate: validEndDate(values.headerStartDate.time, values.headerEndDate.time)
    ? undefined
    : I18n.t('administration.landingPage.header.endDateError')
});