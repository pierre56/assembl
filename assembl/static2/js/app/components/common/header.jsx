// @flow
import * as React from 'react';
import classnames from 'classnames';
import { connect } from 'react-redux';
import { Grid, Row } from 'react-bootstrap';
import { Translate } from 'react-redux-i18n';
import { getIsPhaseCompletedById, getPhaseName } from '../../utils/timeline';
import WhatYouNeedToKnow from '../debate/common/whatYouNeedToKnow';
import HeaderActions from '../debate/common/headerActions';
import { type DebateType } from '../debate/navigation/timelineSegment';

type Props = {
  children?: React.Node,
  title: string,
  subtitle?: string,
  phaseId?: string,
  imgUrl: ?string,
  synthesisTitle?: string,
  additionalHeaderClasses?: string,
  type?: string,
  debate: DebateType,
  timeline: Timeline
};

class Header extends React.Component<Props> {
  render() {
    const { children, title, subtitle, imgUrl, phaseId, synthesisTitle, additionalHeaderClasses, type, timeline } = this.props;
    const { debateData } = this.props.debate;
    let isPhaseCompleted = false;
    let closedPhaseName = '';
    if (phaseId) {
      isPhaseCompleted = getIsPhaseCompletedById(timeline, phaseId);
      if (isPhaseCompleted) {
        closedPhaseName = getPhaseName(timeline, phaseId).toLowerCase();
      }
    }

    const titleClassNames = classnames([additionalHeaderClasses], 'light-title-7');
    return (
      <div className="header-section-container">
        <section className="header-section">
          <Grid fluid className="max-container">
            <div className="header-content">
              <h1 className={titleClassNames}>{title}</h1>
              {subtitle ? <h3 className="light-title-2">{subtitle}</h3> : null}
              {isPhaseCompleted && (
                <h6 className="light-title-4">
                  <Translate value="debate.survey.endPhase" closedPhaseName={closedPhaseName} />
                </h6>
              )}
            </div>
            {type ? <HeaderActions useSocialMedia={debateData.useSocialMedia} type={type} /> : null}
          </Grid>
          <Grid fluid>
            <Row>
              {children}
              <div className="header-bkg" style={imgUrl ? { backgroundImage: `url(${imgUrl})` } : null}>
                &nbsp;
              </div>
              <div className="header-bkg-mask">&nbsp;</div>
            </Row>
          </Grid>
        </section>
        {synthesisTitle && <WhatYouNeedToKnow synthesisTitle={synthesisTitle} />}
      </div>
    );
  }
}

const mapStateToProps = state => ({
  debate: state.debate,
  timeline: state.timeline
});

export default connect(mapStateToProps)(Header);