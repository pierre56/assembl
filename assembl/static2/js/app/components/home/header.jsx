import React from 'react';
import { Translate, I18n } from 'react-redux-i18n';
import { compose, graphql } from 'react-apollo';
import { connect } from 'react-redux';
import { Grid, Row } from 'react-bootstrap';

import Statistic from './header/statistic';
import ParticipateButton from '../common/participateButton';
import { get } from '../../utils/routeMap';
import { getCurrentPhaseData } from '../../utils/timeline';
import { getDiscussionSlug } from '../../utils/globalFunctions';
import { browserHistory } from '../../router';
import manageErrorAndLoading from '../common/manageErrorAndLoading';
import DiscussionQuery from '../../graphql/DiscussionQuery.graphql';

class Header extends React.Component {
  constructor(props) {
    super(props);
    this.displayPhase = this.displayPhase.bind(this);
  }

  displayPhase() {
    const slug = { slug: getDiscussionSlug() };
    const { timeline } = this.props;
    const { currentPhaseIdentifier } = getCurrentPhaseData(timeline);
    browserHistory.push(get('debate', { ...slug, phase: currentPhaseIdentifier }));
  }

  render() {
    const { timeline, data: { discussion } } = this.props;
    const { title, subtitle, headerImage, logoImage, buttonLabel, startDate, endDate } = discussion;
    return (
      <section className="home-section header-section">
        <Grid fluid className="max-container">
          <div className="header-content">
            {logoImage && logoImage.externalUrl ? <img className="header-logo" src={logoImage.externalUrl} alt="logo" /> : null}
            <div className="max-text-width">
              {title ? <h1 className="light-title-1">{title}</h1> : null}
              <h4 className="light-title-4 uppercase margin-m">
                {subtitle ? <span dangerouslySetInnerHTML={{ __html: subtitle }} /> : null}
                {startDate && endDate ? (
                  <div>
                    <Translate
                      value="home.from_start_to_end"
                      start={I18n.l(startDate, { dateFormat: 'date.format' })}
                      end={I18n.l(endDate, { dateFormat: 'date.format' })}
                    />
                  </div>
                ) : null}
              </h4>
              <div className="margin-l">
                <ParticipateButton
                  displayPhase={this.displayPhase}
                  timeline={timeline}
                  btnClass="light"
                  btnLabel={buttonLabel || null}
                />
              </div>
            </div>
          </div>
        </Grid>
        <Grid fluid>
          <Row>
            <Statistic />
            {headerImage && headerImage.externalUrl ? (
              <div className="header-bkg" style={{ backgroundImage: `url(${headerImage.externalUrl})` }}>
                &nbsp;
              </div>
            ) : (
              <div className="header-bkg">&nbsp;</div>
            )}
          </Row>
        </Grid>
      </section>
    );
  }
}

const mapStateToProps = state => ({
  lang: state.i18n.locale,
  timeline: state.timeline
});

export default compose(connect(mapStateToProps), graphql(DiscussionQuery), manageErrorAndLoading({ displayLoader: false }))(
  Header
);