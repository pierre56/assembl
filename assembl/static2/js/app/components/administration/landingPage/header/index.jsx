// @flow
import React from 'react';
import { connect } from 'react-redux';
import { type ApolloClient, graphql, compose, withApollo } from 'react-apollo';
import { I18n } from 'react-redux-i18n';
import { Field } from 'react-final-form';
import setFieldTouched from 'final-form-set-field-touched';
import SectionTitle from '../../../administration/sectionTitle';
import Helper from '../../../common/helper';
import { load, postLoadFormat } from './load';
import { save, createMutationsPromises } from './save';
import LoadSaveReinitializeForm from '../../../form/LoadSaveReinitializeForm';
import FileUploaderFieldAdapter from '../../../form/fileUploaderFieldAdapter';
import DatePickerFieldAdapter from '../../../form/datePickerFieldAdapter';
import MultilingualTextFieldAdapter from '../../../form/multilingualTextFieldAdapter';
import MultilingualRichTextFieldAdapter from '../../../form/multilingualRichTextFieldAdapter';
import AdminForm from '../../../form/adminForm';
import Loader from '../../../common/loader';
import { type DateTime } from '../../../form/types.flow';
import { validStartDate, validEndDate, validateDatePicker } from './validate';
import DiscussionQuery from '../../../../graphql/DiscussionQuery.graphql';
import { convertISO8601StringToDate } from '../../../../utils/globalFunctions';
import manageErrorAndLoading from '../../../common/manageErrorAndLoading';

type Props = {
  client: ApolloClient,
  editLocale: string,
  data: {
    discussion: {
      startDate: string,
      endDate: string
    }
  }
};

type State = {
  startDate: DateTime,
  endDate: DateTime,
  startDateConflict: boolean,
  endDateConflict: boolean
};

const loading = <Loader />;

export class DumbCustomizeHeader extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
    const { startDate, endDate } = props.data.discussion;
    this.state = {
      startDate: startDate ? convertISO8601StringToDate(startDate) : null,
      endDate: endDate ? convertISO8601StringToDate(endDate) : null,
      startDateConflict: false,
      endDateConflict: false
    };
  }

  onStartChange = (e: DateTime) => {
    const { endDate } = this.state;
    let startDateConflict = false;
    if (!validStartDate(e, endDate)) {
      startDateConflict = true;
    } else {
      startDateConflict = false;
    }
    this.setState(() => ({
      startDate: e,
      startDateConflict: startDateConflict
    }));
  };

  onEndChange = (e: DateTime) => {
    const { startDate } = this.state;
    let endDateConflict = false;
    if (!validEndDate(startDate, e)) {
      endDateConflict = true;
    } else {
      endDateConflict = false;
    }
    this.setState(() => ({
      endDate: e,
      endDateConflict: endDateConflict
    }));
  };

  render() {
    const { client, editLocale } = this.props;
    const { startDateConflict, endDateConflict } = this.state;
    return (
      <div className="admin-box">
        <SectionTitle
          title={I18n.t('administration.landingPage.header.title')}
          annotation={I18n.t('administration.annotation')}
        />
        <LoadSaveReinitializeForm
          load={fetchPolicy => load(client, editLocale, fetchPolicy)}
          loading={loading}
          postLoadFormat={postLoadFormat}
          createMutationsPromises={createMutationsPromises(client)}
          save={save}
          mutators={{ setFieldTouched: setFieldTouched }}
          validate={validateDatePicker}
          render={({ handleSubmit, pristine, submitting, form }) => (
            <div className="admin-content">
              <AdminForm
                handleSubmit={handleSubmit}
                pristine={pristine || startDateConflict || endDateConflict}
                submitting={submitting}
              >
                <div className="form-container">
                  <Helper
                    classname="admin-paragraph"
                    label={I18n.t('administration.landingPage.header.helper')}
                    helperUrl={'/static2/img/helpers/landing_page_admin/header.png'}
                    helperText={I18n.t('administration.helpers.landingPage.header')}
                  />
                  <Field
                    editLocale={editLocale}
                    name="headerTitle"
                    component={MultilingualTextFieldAdapter}
                    label={I18n.t('administration.landingPage.header.titleLabel')}
                    required
                  />
                  <Field
                    editLocale={editLocale}
                    withAttachmentButton={false}
                    name="headerSubtitle"
                    component={MultilingualRichTextFieldAdapter}
                    label={I18n.t('administration.landingPage.header.subtitleLabel')}
                    required
                  />
                  <Field
                    editLocale={editLocale}
                    name="headerButtonLabel"
                    component={MultilingualTextFieldAdapter}
                    label={I18n.t('administration.landingPage.header.buttonLabel')}
                    required={false}
                  />
                  <Field
                    name="headerImage"
                    component={FileUploaderFieldAdapter}
                    label={I18n.t('administration.landingPage.header.headerImage')}
                  />
                  <Field
                    name="headerLogoImage"
                    component={FileUploaderFieldAdapter}
                    label={I18n.t('administration.landingPage.header.logoDescription')}
                  />
                  <div className="section-description">{I18n.t('administration.landingPage.header.dateDescription')}</div>
                  <Field
                    name="headerStartDate"
                    component={DatePickerFieldAdapter}
                    picker={{ pickerType: I18n.t('administration.landingPage.header.startDate') }}
                    editLocale={editLocale}
                    placeHolder={I18n.t('administration.landingPage.header.timePlaceholder')}
                    showTime={false}
                    hasConflictingDate={this.state.startDateConflict}
                    onDateChange={this.onStartChange}
                    form={form}
                  />
                  <Field
                    name="headerEndDate"
                    component={DatePickerFieldAdapter}
                    picker={{ pickerType: I18n.t('administration.landingPage.header.endDate') }}
                    editLocale={editLocale}
                    placeHolder={I18n.t('administration.landingPage.header.timePlaceholder')}
                    showTime={false}
                    hasConflictingDate={this.state.endDateConflict}
                    onDateChange={this.onEndChange}
                    form={form}
                  />
                </div>
              </AdminForm>
            </div>
          )}
        />
      </div>
    );
  }
}

const mapStateToProps = ({ admin: { editLocale } }) => ({
  editLocale: editLocale
});

export default compose(
  connect(mapStateToProps),
  graphql(DiscussionQuery, {
    options: ({ editLocale }) => ({
      variables: {
        lang: editLocale,
        nextView: null
      }
    })
  }),
  manageErrorAndLoading({ displayLoader: true }),
  withApollo
)(DumbCustomizeHeader);