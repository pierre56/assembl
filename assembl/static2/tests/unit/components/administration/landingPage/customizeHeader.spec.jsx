import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import { DumbCustomizeHeader } from '../../../../../js/app/components/administration/landingPage/header/index';

describe('ManageModules component', () => {
  it('should render a page to manage the landing page header', () => {
    const data = {
      discussion: {
        logoImage: null,
        headerImage: {
          mimeType: 'image/png',
          title: 'Screen Shot 2018-11-23 at 5.09.00 PM.png',
          externalUrl: 'http://localhost:6543/data/Discussion/32/documents/689/data'
        },
        title: 'Title in EN 13',
        titleEntries: [
          {
            localeCode: 'en',
            value: 'Title in EN 13',
            __typename: 'LangStringEntry'
          }
        ],
        subtitle: '<p>subtitle in EN 13</p>',
        subtitleEntries: [
          {
            localeCode: 'en',
            value: '<p>subtitle in EN 13</p>',
            __typename: 'LangStringEntry'
          }
        ],
        buttonLabel: 'button in EN 13',
        buttonLabelEntries: [
          {
            localeCode: 'en',
            value: 'button in EN 13',
            __typename: 'LangStringEntry'
          }
        ],
        homepageUrl: null,
        loginData: {
          local: true,
          url: '/qa-debate-fred/login'
        },
        startDate: '2017-09-13T00:00:00+00:00',
        endDate: '2017-11-25T00:00:00+00:00'
      }
    };
    const props = {
      client: null,
      editlocale: 'en',
      data: { ...data }
    };

    const shallowRenderer = new ShallowRenderer();
    shallowRenderer.render(<DumbCustomizeHeader {...props} />);
    const result = shallowRenderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });
});