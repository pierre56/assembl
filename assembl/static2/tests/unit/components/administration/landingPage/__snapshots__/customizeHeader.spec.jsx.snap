// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManageModules component should render a page to manage the landing page header 1`] = `
<div
  className="admin-box"
>
  <SectionTitle
    annotation="* Fields are required."
    title="Header"
  />
  <LoadSaveReinitializeForm
    createMutationsPromises={[Function]}
    load={[Function]}
    loading={
      <Loader
        textHidden={false}
      />
    }
    mutators={
      Object {
        "setFieldTouched": [Function],
      }
    }
    postLoadFormat={[Function]}
    render={[Function]}
    save={[Function]}
    validate={[Function]}
  />
</div>
`;
